sql/src/main/java/io/crate/planner/operators/Join.java

```diff
@@ -463,4 +498,38 @@ public long numExpectedRows() {
     public <C, R> R accept(LogicalPlanVisitor<C, R> visitor, C context) {
         return visitor.visitJoin(this, context);
     }
+
+    @Override
+    public LogicalPlan tryOptimize(@Nullable LogicalPlan pushDown) {
+        if (pushDown instanceof Order) {
+            /* Move the orderBy expression to the sub-relation if possible.
+             *
+             * This is possible because a nested loop preserves the ordering of the input-relation
+             * IF:
+             *   - the order by expressions only operate using fields from a single relation
+             *   - that relation happens to be on the left-side of the join
+             *   - there is no outer join involved in the whole join (outer joins may create null rows - breaking the ordering)
+             */
+            if (hasOuterJoins == false) {
+                Set<AnalyzedRelation> relationsInOrderBy =
+                    Collections.newSetFromMap(new IdentityHashMap<AnalyzedRelation, Boolean>());
```

Comment:
Could this be a simple HashSet? (Fields implement equals and hashcode already)

Reply:
This code is copied and i think the identity hash map is needed in the case of self joins (same table with diff alias)